### DISCUSSION TITLE
Confusion about how to generate output file when there is a custom scalar type

### DISCUSSION THREAD

**Discussion by oveddan** [0no-co/gql.tada#412] [Q&A]
I have a schema with custom scalar type `BigInt`:
```
type TransactionInfo @entity {
  id: ID!
  chainId: BigInt!
  network: String!
  block: BigInt!
  logIndex: BigInt!
  address: Bytes!
  timestamp: BigInt!
}
```

When I try to run `gql-tada generate output` with this schema, I get an error "Error: Unknown type: "BigInt"."

How can I instruct, in the generate output step, how to handle these custom scalars? I see there is a section on custom scalars, but that seems to only apply after the output is generated?

**Comment by kitten** [comment]
Do you actually have a `scalar BigInt` definition in your SDL file? That error sounds more like the schema parsing has failed 😅

**Reply by oveddan** [reply]
We're currently using the [Graph Protocol's tooling](https://github.com/graphprotocol/graph-tooling/tree/main/packages/cli) to generate types from the schema.  It supports their custom scalar types [explained here](https://thegraph.com/docs/en/developing/creating-a-subgraph/#built-in-scalar-types).

We use the command `graph codegen` and it works.

I'm not quite sure the internals of how they set it up to parse this custom scalar type.  Do you have any idea of how to integrate with `gql.data`?

For reference, `graph codegen` task is able to successfully generate a `schema.ts` file.

The package where we are attempting this [is here.](https://github.com/ourzora/zora-protocol/tree/main/packages/creator-subgraph)

**Reply by JoviDeCroock** [reply]
Hmm, that would actually imply that the schema you have is invalid as it is missing a `scalar BigInt` definition. 😅 

You could manually pull down the schema by doing `gql-tada generate schema` and then adding the `scalar BigInt` yourself.

That being said with a valid schema you would just have to add your `scalar` to your `initGraphQLTada` config and we would translate it to the right type. https://gql-tada.0no.co/reference/gql-tada-api#initgraphqltada

**Reply by oveddan** [reply]
Thanks for the quick reply.  It seems like a bit of a chicken and egg issue - I can't run `pnpm gql-tada generate schema ./schema.graphql` - when I do I get the same error `Error: Unknown type: "BigInt".` because the scalar type is not defined in the schema.  I believe the graph protocol's tooling injects that custom scalar type dynamically when doing codegen.

And correspondingly, I cannot setup `initGraphQLTada` because it requires the introspection object from `graphql-env.d.ts` which has not been generated yet.

**Reply by oveddan** [reply]
It would be great if you could specify custom scalar types at runtime when generating the output file.

**Reply by kitten** [reply]
Can you look into whether the schema that you get from the Graph is valid please?

If you have an SDL file, or if you pull one using `generate schema`:
- does it contain `scalar BigInt` as one line?
- if not, does it reference `BigInt` as in the example you've posted?
- does it parse with [`buildSchema`](https://github.com/expo/expo-easy-poc/commit/ce3c0a1567053254dc3ff5e9024b82b6397820e1)?

There's some other notes in other discussions about this, but we don't accept incomplete SDL (i.e. no modifications or split up SDL files will be pulled in) because we're not making a schema builder, and there's too many ways to build schemas. In other words, SDL/introspection query results are basically a standard way of conveying a complete schema, so we don't want to instead also support all the ways of building a schema "from scratch" 😅 

In your case it's of course presumably not built from scratch, but I do want to get to the bottom of this. If the schema is invalid, we can maybe add a workaround and automatically add scalars for missing output and input types (although, I have to think about this further, this might be dangerous and confusing)

But primarily it sounds to me like a third-party bug, so it might be better to confirm the above and report it to The Graph
